content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+
+## A Region-Based type extension for Bril
+
+This work, based on the [cyclone] language, seeks to bring region-based memory management to the Bril [memory extension][memex]. Each pointer to memory exists within the static scope of assigned memory called a region. Regions are hierarchical such that they can be nested within each other. Data that is allocated inside of a region can only be used within that region or inner-nested regions. When the scope of a region ends, all of the allocations that belong to that region are freed. In this way, the compiler statically rules out the previously stated, common memory bugs by enforcing when and for how long an allocation can be used for. For Bril, this comes in the form of an extension to the Bril Json representation and a static analysis tool that accepts this extension of Bril and converts it to standard Bril with manual memory management inserted into the program.
+
+## Extension Design
+
+The Type enum in the [bril_rs] library is modified as follows:
+
+```rust
+// I've removed the parts of this enum used in serializing/deserializing to json.
+pub enum Type {
+    Int,
+    Bool,
+    Float,
+    Pointer(Box<Type>),
+
+    /// New addition
+    PointerRegions {
+        pointer_type : Box<Type>,
+        ownership : Option<Ownership>,
+        region: String,
+    },
+    ///
+}
+```
+
+The new addition is a pointer type which is annotated with the region it exists in and a concept of whether a pointer owns the memory it is pointing to. In Bril, every ```id```, ```load```, ```alloc```, and ```ptr_add``` can create a new pointer into memory. However, not all pointers are created equal. Pointer's
+created by ```alloc``` are the original pointers into the zeroth index of memory and are treated like the "Owners" of that memory. It is these pointers that need to be freed when the scope of their region ends. Pointers created by ```id```, ```load```, and ```ptr_add``` are merely memory borrowers with views into heap memory. They still need to be statically checked that they aren't being used after the region scope ends but do not need a corresponding ```free``` instruction like memory owners do. What is important about this typing is that both pointer types can exist within a program depending on whether you want the region to be inferred or explicitly stated. Whether a pointer owns its memory is trivial to decide and it is not expected that programmers included this annotation. As will be touched on in the evaluation section, this opt-in use of explicit regions makes converting valid programs to this extension simple.
```

Comment:
I'll add a sentence about passing a pointer into a function. I treat this as an implicit `id` call so a function can only take a pointer that borrows memory. A pointer returned from a function is always owned specifically to avoid having to do an interprocedural analysis. This seemed like a reasonable limitation/tradeoff.

I think this is where some of my constraints come into play. Having an allocated pointer dominate all of its exit's and not allowing a reassignment to that variable should avoid this.

The last two are similar. I'm not destroying any pointers. Any time you are creating a new pointer from an old one, the new one will be a borrow. In this way, you can't store an "owning" pointer into memory because we are not destroying the original pointer. I think if I allowed that I would need some kind of runtime mechanism/destructor every time I want to free memory that can contain pointers. I don't want to "destroy" pointers.